cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -0,0 +1,532 @@
+  }
+  void TearDown() override {
+    auto containers = gen2Client_->ListFileSystems();
+    for (auto c : containers.FileSystems) {
+      auto fileSystemClient = gen2Client_->GetFileSystemClient(c.Name);
+      fileSystemClient.DeleteIfExists();
+    }
+  }
+  void AssertObjectContents(
+      Azure::Storage::Files::DataLake::DataLakeServiceClient* client,
+      const std::string& container, const std::string& path_to_file,
+      const std::string& expected) {
+    auto pathClient_ =
+        std::make_shared<Azure::Storage::Files::DataLake::DataLakePathClient>(
+            client->GetUrl() + container + "/" + path_to_file,
+            options_.storage_credentials_provider);
+    auto size = pathClient_->GetProperties().Value.FileSize;
+    if (size == 0) {
+      return;
+    }
+    auto buf = AllocateResizableBuffer(size, fs_->io_context().pool());
+    Azure::Storage::Blobs::DownloadBlobToOptions downloadOptions;
+    Azure::Core::Http::HttpRange range;
+    range.Offset = 0;
+    range.Length = size;
+    downloadOptions.Range = Azure::Nullable<Azure::Core::Http::HttpRange>(range);
+    auto fileClient_ =
+        std::make_shared<Azure::Storage::Files::DataLake::DataLakeFileClient>(
+            client->GetUrl() + container + "/" + path_to_file,
+            options_.storage_credentials_provider);
+    auto result = fileClient_
+                      ->DownloadTo(reinterpret_cast<uint8_t*>(buf->get()->mutable_data()),
+                                   size, downloadOptions)
+                      .Value;
+    buf->get()->Equals(
+        Buffer(const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(&expected[0])),
+               expected.size()));
+  }
+};
+
+TEST_F(TestAzureFileSystem, CreateDir) {
+  // New container
+  AssertFileInfo(fs_.get(), "container3", FileType::NotFound);
+  ASSERT_OK(fs_->CreateDir("container3"));
+  AssertFileInfo(fs_.get(), "container3", FileType::Directory);
+
+  // Existing container
+  ASSERT_OK(fs_->CreateDir("container"));
+  AssertFileInfo(fs_.get(), "container", FileType::Directory);
+
+  ASSERT_RAISES(IOError, fs_->CreateDir(""));
+
+  // Existing "file", should fail
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/somefile"));
+
+  // directory, false
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/newdir/newsub/newsubsub", false));
+
+  // directory, true
+  ASSERT_RAISES(IOError, fs_->CreateDir("container/newdir/newsub/newsubsub", true));
+}
+
+TEST_F(TestAzureFileSystem, DeleteDir) {
+  // Container
+  ASSERT_OK(fs_->DeleteDir("container"));
+  AssertFileInfo(fs_.get(), "container", FileType::NotFound);
+
+  // Nonexistent Container
+  ASSERT_OK(fs_->DeleteDir("container3"));
+  AssertFileInfo(fs_.get(), "container3", FileType::NotFound);
+
+  // root
+  ASSERT_RAISES(NotImplemented, fs_->DeleteDir(""));
+
+  // C/F
+  ASSERT_RAISES(IOError, fs_->DeleteDir("container/somefile"));
+
+  // C/NF
+  ASSERT_RAISES(IOError, fs_->DeleteDir("container/somefile19"));
+
+  // C/ND/D
```

Comment:
Does "ND" mean a nonexistent directory and "D" mean an existent directory?
Why an existent directory under a nonexistent directory?